src/ngx_http_lua_output.c

```diff
@@ -802,4 +802,105 @@ ngx_http_lua_flush_cleanup(void *data)
     ctx->flushing_coros--;
 }
 
+
+#ifndef NGX_LUA_NO_FFI_API
+int
+ngx_http_lua_ffi_write(ngx_http_request_t *r, const char *buf, size_t offset,
+                       size_t len, char **err)
+{
+    ngx_http_lua_ctx_t          *ctx;
+    ngx_uint_t                   size;
+    ngx_buf_t                   *b;
+    ngx_chain_t                 *cl;
+    ngx_int_t                    rc;
+    const char                  *str_sub;
+
+    if (r == NULL) {
+        *err = "no request object found";
+        return NGX_ERROR;
+    }
+
+    if (buf == NULL) {
+        *err = "no buffer found";
+        return NGX_ERROR;
+    }
+
+    size = ngx_strlen(buf);
```

Comment:
@hkmo Nope, the first argument of `ffi_write` is defined to be a cdata object of the type `const char *` which represents the starting pointer of a C buffer. You can directly feed a Lua string there just because LuaJIT can do the implicit Lua string -> `const char *` cdata casting. The `ffi_` prefix means that this API is unsafe per se so it's the caller's responsibility to ensure the validity of all the function argument values.